### PR TITLE
feat(client): extended where

### PR DESCRIPTION
**Pull request by millsp**
Integration PR for https://github.com/prisma/prisma-engines/pull/3281

Tests will pass once the engine has been merged

### CODE REVIEW

File: packages/client/src/generation/TSClient/common.ts [hunk 1]
```diff
@@ -416,6 +416,16 @@ type _Record<K extends keyof any, T> = {
   [P in K]: T;
 };
 
+// cause typescript not to expand types and preserve names
+type NoExpand<T> = T extends unknown ? T : never;
```
Comment (SevInf) [inline packages/client/src/generation/TSClient/common.ts:420]: For my own understanding, when `T extends unknown` would not be true?

File: packages/client/src/generation/TSClient/common.ts [hunk 2]
```diff
@@ -416,6 +416,16 @@ type _Record<K extends keyof any, T> = {
...
+
+// this type assumes the passed object is entirely optional
+type AtLeast<O extends object, K extends string> = NoExpand<
```
Comment (SevInf) [inline packages/client/src/generation/TSClient/common.ts:423]: What would you think about pulling this type into separate file and writing a couple of test for it using `expectTypeOf` ?

File: packages/client/tests/functional/extended-where/findFirstOrThrow.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  })
+
+  test('findFirstOrThrow with cursor 1 unique (PK)', async () => {
+    const data = await prisma.post.findFirstOrThrow({
+      cursor: {
+        id: vars.postId2,
+      },
+    })
+
+    expect(data?.id).toBe(vars.postId2)
```
Comment (SevInf) [inline packages/client/tests/functional/extended-where/findFirstOrThrow.ts:21]:
```suggestion
    expect(data.id).toBe(vars.postId2)
```

`data` is non-nullabale in this case

File: packages/client/tests/functional/extended-where/findFirstOrThrow.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+  test('findFirstOrThrow with cursor 2 uniques (PK & non-PK)', async () => {
+    const data = await prisma.post.findFirstOrThrow({
+      cursor: {
+        id: vars.postId2,
+        title: 'Hello World 2',
+      },
+    })
+
+    expect(data?.id).toBe(vars.postId2)
```
Comment (SevInf) [inline packages/client/tests/functional/extended-where/findFirstOrThrow.ts:32]:
```suggestion
    expect(data.id).toBe(vars.postId2)
```

File: packages/client/tests/functional/extended-where/findFirstOrThrow.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  })
+
+  test('findFirstOrThrow with cursor 1 unique (non-PK)', async () => {
+    const data = await prisma.post.findFirstOrThrow({
+      cursor: {
+        title: 'Hello World 2',
+      },
+    })
+
+    expect(data?.id).toBe(vars.postId2)
```
Comment (SevInf) [inline packages/client/tests/functional/extended-where/findFirstOrThrow.ts:42]:
```suggestion
    expect(data.id).toBe(vars.postId2)
```

File: packages/client/tests/functional/extended-where/findUnique.ts [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+declare let prisma: PrismaClient
+
+let vars: Awaited<ReturnType<typeof setup>>
+testMatrix.setupTestSuite(() => {
+  beforeAll(async () => {
+    vars = await setup(prisma)
+  })
+
+  test('findUnique with where 1 unique (PK)', async () => {
+    await prisma.user.findUnique({
```
Comment (SevInf) [inline packages/client/tests/functional/extended-where/findUnique.ts:15]: Shouldn't we assert something here?

**Review by SevInf — APPROVED** [review]
Couple of comments/questions, nothing blocking

File: packages/client/src/generation/TSClient/common.ts (hunk 1 above)
Comment (millsp) [reply packages/client/src/generation/TSClient/common.ts:420]: It's always true

File: packages/client/tests/functional/extended-where/findUnique.ts (hunk 6 above)
Comment (millsp) [reply packages/client/tests/functional/extended-where/findUnique.ts:15]: Yes, I think I should assert the results

File: packages/client/src/generation/TSClient/common.ts (hunk 1 above)
Comment (millsp) [reply packages/client/src/generation/TSClient/common.ts:420]: It's completely artificial. TypeScript has a way (I cannot recall how they call this) to preserve type names and not expand them when wrapped inside more complex generic types. So here, i am improving DX by utilizing that behavior at our advantage. See with and without:

Without:
Screenshot from 2022-10-17 06-00-57

With:
Screenshot from 2022-10-17 06-00-38

Comment (SevInf) [reply packages/client/src/generation/TSClient/common.ts:420]: Thanks for the explanations!

Comment (millsp) [reply packages/client/src/generation/TSClient/common.ts:420]: Auto completion stays the same within the object.

**Pull request merged**